Prevent makers from reading offer list during `makerExecute`

espendk opened the pull request:
This PR prevents makers from reading the offer list during `makerExecute`.

This is done for two reasons:
1. To prevent `makerExecute` from detecting cleaning
2. To avoid the gas cost of keeping the offer list up-to-date for `makerExecute`. A number of optimizations are in place which mean that `local`, leafs, and levels may not be up-to-date.

Two measures are implemented:
- The existing offer list lock is now enforced as a read-lock as well in all the view functions.
- All fields in the `SingleOrder` struct that pertain to the offer list/tick tree structure are cleared (set to zero) before it is passed to `makerExecute`.


Closes https://github.com/mangrovedao/mangrove-core/issues/503

espendk committed:
feat: Prevent read of locked offer lists

espendk committed:
refactor(test): Make posthook reentrancy tests check that posthook was called

espendk committed:
feat: Hide offer list fields from makerExecute

espendk committed:
fix(test): Ensure callback with assertions is called in makerExecute test

adhusson commented on preprocessing/structs.ts:
```diff
@@ -118,6 +118,9 @@ import "mgv_lib/LogPriceConversionLib.sol";
 using OfferUnpackedExtra for OfferUnpacked global;
 
+// cleanup-mask: 0s at location of fields to hide from maker, 1s elsewhere
+uint constant hide_fields_from_maker_mask = ~(prev_mask_inv | next_mask_inv);
```
```suggestion
uint constant HIDE_FIELDS_FROM_MAKER_MASK = ~(prev_mask_inv | next_mask_inv);
```

espendk replied:
Addressed in 20bcff1575e55fdf83fbd3b570d066cfb589d250

adhusson commented on preprocessing/structs.ts:
```diff
@@ -133,6 +136,13 @@ library OfferPackedExtra {
+    unchecked {
+      return OfferPacked.wrap(
+        OfferPacked.unwrap(offer)
+        & hide_fields_from_maker_mask);
```
```suggestion
        & HIDE_FIELDS_FROM_MAKER_MASK);
```

adhusson commented on preprocessing/structs.ts:
```diff
@@ -304,6 +314,9 @@ import {Density, DensityLib} from "mgv_lib/DensityLib.sol";
 using LocalUnpackedExtra for LocalUnpacked global;
 
+// cleanup-mask: 0s at location of fields to hide from maker, 1s elsewhere
+uint constant hide_fields_from_maker_mask = ~(tickPosInLeaf_mask_inv | level0_mask_inv | level1_mask_inv | level2_mask_inv | last_mask_inv);
```
```suggestion
uint constant HIDE_FIELDS_FROM_MAKER_MASK = ~(tickPosInLeaf_mask_inv | level0_mask_inv | level1_mask_inv | level2_mask_inv | last_mask_inv);
```

adhusson commented on preprocessing/structs.ts:
```diff
@@ -317,6 +330,13 @@ library LocalPackedExtra {
+    unchecked {
+      return LocalPacked.wrap(
+        LocalPacked.unwrap(local)
+        & hide_fields_from_maker_mask);
```
```suggestion
        & HIDE_FIELDS_FROM_MAKER_MASK);
```

adhusson commented on src/MgvView.sol:
```diff
@@ -60,6 +68,7 @@ contract MgvView is MgvCommon {
   {
     unchecked {
       (MgvStructs.GlobalPacked __global, MgvStructs.LocalPacked __local) = config(olKey);
+      unlockedMarketOnly(__local);
```
Can a contract still access the global Mangrove config when at least one market is locked?

espendk replied:
Yes, they can access it via another offer list. The test `test_offer_list_read_on_reentrancy_succeeds` checks this: https://github.com/mangrovedao/mangrove-core/pull/512/files#diff-dd075adeff468191afc500c88722b7198985b2c71809811e9272fdfed903d160R739-R762

Do you think we should have a view function for reading only the global config?

adhusson replied:
yes -- definitely violates the least surprise principles otherwise

espendk replied:
Addressed in 3aa854b8d2a4e829c8ffc6a7b21d814a358fbd56

adhusson commented on src/preprocessed/MgvLocal.post.sol:
```diff
@@ -39,6 +39,9 @@ import {Density, DensityLib} from "mgv_lib/DensityLib.sol";
 using LocalUnpackedExtra for LocalUnpacked global;
 
+// cleanup-mask: 0s at location of fields to hide from maker, 1s elsewhere
+uint constant hide_fields_from_maker_mask = ~(tickPosInLeaf_mask_inv | level0_mask_inv | level1_mask_inv | level2_mask_inv | last_mask_inv);
```
```suggestion
uint constant HIDE_FIELDS_FROM_MAKER_MASK = ~(tickPosInLeaf_mask_inv | level0_mask_inv | level1_mask_inv | level2_mask_inv | last_mask_inv);
```

adhusson commented on src/preprocessed/MgvLocal.post.sol:
```diff
@@ -52,6 +55,13 @@ library LocalPackedExtra {
+    unchecked {
+      return LocalPacked.wrap(
+        LocalPacked.unwrap(local)
+        & hide_fields_from_maker_mask);
```
```suggestion
        & HIDE_FIELDS_FROM_MAKER_MASK);
```

adhusson commented on src/preprocessed/MgvOffer.post.sol:
```diff
@@ -33,6 +33,9 @@ import "mgv_lib/LogPriceConversionLib.sol";
 using OfferUnpackedExtra for OfferUnpacked global;
 
+// cleanup-mask: 0s at location of fields to hide from maker, 1s elsewhere
+uint constant hide_fields_from_maker_mask = ~(prev_mask_inv | next_mask_inv);
```
```suggestion
uint constant HIDE_FIELDS_FROM_MAKER_MASK = ~(prev_mask_inv | next_mask_inv);
```

adhusson commented on src/preprocessed/MgvOffer.post.sol:
```diff
@@ -48,6 +51,13 @@ library OfferPackedExtra {
+    unchecked {
+      return OfferPacked.wrap(
+        OfferPacked.unwrap(offer)
+        & hide_fields_from_maker_mask);
```
```suggestion
        & HIDE_FIELDS_FROM_MAKER_MASK);
```

adhusson reviewed (commented):
Clean :) Do we have a sense of the gas cost change?

espendk committed:
refactor: Rename hide_fields_from_maker_mask to upper case

espendk commented:
> Clean :) Do we have a sense of the gas cost change?

624 gas per offer taken w/o optimization
115 gas per offer taken with optimization

Determined with @lnist 's gas test suite :-) Which we should really extend to include diff functionality similar to `forge snapshot --diff`!

lnist reviewed (approved):
The split to have MgvView shows its brilliance here.

espendk committed:
perf: Save gas by clearing fields in `makerExecute` iso. `execute`

Clearing offer and local fields in `makerExecute` means we don't have to save and restore the old values in `execute`.

espendk committed:
Revert "perf: Save gas by clearing fields in `makerExecute` iso. `execute`"

This reverts commit 470369bca6910d95f28928d071e67e3de930cdc0.

espendk committed:
feat: Add view functions for the global config

This allows easy reading of the global config even if an offer list is locked.

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-read-lock

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-read-lock

espendk committed:
docs: Add comment on why field clearing happens in `execute`
